### DISCUSSION TITLE
How do I search with DataType.BINARY_VECTOR

### DISCUSSION THREAD

**Discussion by WolfNiu** [milvus-io/milvus#33986] [Q&A and General discussion] [answered]
Hi, I have the following code that computes Hamming distance with DataType.BINARY_VECTOR. However, the last step "client.search" gives me an error when I try to search with binary vectors.

```
from pymilvus import MilvusClient, DataType
from pathlib import Path
import numpy as np


DB_FILE = 'demo.db'
DIM = 4096
COLLECTION_NAME = 'dim_reduction'
METRIC_TYPE = 'HAMMING'
INDEX_TYPE = 'BIN_FLAT'
DATATYPE = DataType.BINARY_VECTOR
DTYPE = np.bool_


# Remove DB_FILE if exists
db_path = Path(DB_FILE)
if db_path.exists():
    db_path.unlink()

# Build client
client = MilvusClient(DB_FILE)

# Create schema
schema = MilvusClient.create_schema()
schema.add_field(field_name="id", datatype=DataType.INT64, is_primary=True)
schema.add_field(field_name='text', datatype=DataType.VARCHAR, max_length=1024)
schema.add_field(field_name='vector', datatype=DATATYPE, dim=DIM)

# Create collection
client.create_collection(
    collection_name=COLLECTION_NAME,
    schema=schema,
)

# Insert data
data = [
    {'id': i, 'vector': np.array([1] * DIM, dtype=DTYPE), 'text': f'doc {i}'}
    for i in range(100)
]
client.insert(collection_name=COLLECTION_NAME, data=data)

# Create index
index_params = MilvusClient.prepare_index_params()
index_params.add_index(
    field_name='vector',
    metric_type=METRIC_TYPE,
    index_type=INDEX_TYPE,
)
client.create_index(
    collection_name=COLLECTION_NAME,
    index_params=index_params,
)

# Search
search_params = {
    'metric_type': METRIC_TYPE,
    'params': {},
}

result = client.search(
    collection_name=COLLECTION_NAME,
    data=[np.array([1] * DIM, dtype=DTYPE)],
    limit=2,
    search_params=search_params,
)
print(result)
```

My question is, how to specify "data" to search in this case? Thank you in advance.

**Comment by yhmo** [comment, marked as answer]
Binary vector should be a byte array.
The example of pymilvus shows how to provide a binary vector to milvus: https://github.com/milvus-io/pymilvus/blob/f7a4839a8a6b05620985d25cde47b63247a561e7/examples/binary_example.py#L23

**Reply by WolfNiu** [reply]
Thank you yhmo! 
I see that the .db file created with binary vector is only 25% of the storage taken by the same data but with float32. I wonder why that's the case -- I thought binary vector should be only 1-bit, and the storage will be 1/32 that of the float32.

**Reply by yhmo** [reply]
Raw vector data in Milvus is stored in parquet format, compression is enabled. The actual size depends on the compression ratio of different data(bytes, floats might get different compression ratio).
